pkg/occlient/projects.go

```diff
@@ -0,0 +1,236 @@
+					case "":
+						if val.Type == watch.Deleted {
+							projectChannel <- projectStatus
+							break
+						}
+						if val.Type == watch.Error {
+							watchErrorChannel <- errors.Errorf("failed watching the deletion of project %s", name)
+							break
+						}
+					}
+
+				} else {
+					watchErrorChannel <- errors.New("unable to convert event object to Project")
+					break
+				}
+
+			}
+			close(projectChannel)
+			close(watchErrorChannel)
+		}()
+
+		select {
+		case val := <-projectChannel:
+			klog.V(3).Infof("Deletion information for project: %+v", val)
+			return nil
+		case err := <-watchErrorChannel:
+			return err
+		case <-time.After(waitForProjectDeletionTimeOut):
+			return errors.Errorf("waited %s but couldn't delete project %s in time", waitForProjectDeletionTimeOut, name)
+		}
+
+	}
+
+	// Return nil since we don't bother checking for the watcher..
+	return nil
+}
+
+// CreateNewProject creates project with given projectName
+func (c *Client) CreateNewProject(projectName string, wait bool) error {
+	// Instantiate watcher before requesting new project
+	// If watched is created after the project it can lead to situation when the project is created before the watcher.
+	// When this happens, it gets stuck waiting for event that already happened.
+	var watcher watch.Interface
+	var err error
+	if wait {
+		watcher, err = c.projectClient.Projects().Watch(metav1.ListOptions{
+			FieldSelector: fields.Set{"metadata.name": projectName}.AsSelector().String(),
+		})
+		if err != nil {
+			return errors.Wrapf(err, "unable to watch new project %s creation", projectName)
+		}
+		defer watcher.Stop()
+	}
+
+	projectRequest := &projectv1.ProjectRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: projectName,
+		},
+	}
+	_, err = c.projectClient.ProjectRequests().Create(projectRequest)
+	if err != nil {
+		return errors.Wrapf(err, "unable to create new project %s", projectName)
+	}
+
+	if watcher != nil {
+		for {
+			val, ok := <-watcher.ResultChan()
+			if !ok {
+				break
+			}
+			if prj, ok := val.Object.(*projectv1.Project); ok {
+				klog.V(3).Infof("Status of creation of project %s is %s", prj.Name, prj.Status.Phase)
+				switch prj.Status.Phase {
+				//prj.Status.Phase can only be "Terminating" or "Active" or ""
+				case corev1.NamespaceActive:
+					if val.Type == watch.Added {
+						klog.V(3).Infof("Project %s now exists", prj.Name)
+						return nil
+					}
+					if val.Type == watch.Error {
+						return fmt.Errorf("failed watching the deletion of project %s", prj.Name)
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// IsProjectSupported checks if Project resource type is present on the cluster
+func (c *Client) IsProjectSupported() (bool, error) {
+	return c.isResourceSupported("project.openshift.io", "v1", "projects")
+}
+
+// GetCurrentProjectName returns the current project name
+func (c *Client) GetCurrentProjectName() string {
+	return c.Namespace
```

Comment:
Do we really need a function for this? Is there some design pattern we're following here?

Reply:
I copied it from the original code but I feel it helps in hiding the implementation details.

Reply:
Agree with your reasoning. It's just that I don't really find it very valuable. More so because it's being called at only two places in code and one place in tests:
```sh
$ grep -r "GetCurrentProjectName" --exclude-dir=vendor
Binary file odo matches
pkg/occlient/occlient.go:func (c *Client) GetCurrentProjectName() string {
pkg/occlient/occlient.go:	currentProjectName := c.GetCurrentProjectName()
pkg/occlient/occlient_test.go:				t.Errorf("GetImageStream() = %#v, want %#v and the current project name is %s\n\n", got, tt, fkclient.GetCurrentProjectName())
pkg/catalog/catalog.go:	currentNamespace := client.GetCurrentProjectName()
```

Reply:
But it might a good idea not to leak the logic of fetching the current project in the layers above like catalog.go. The client should be responsible and should hide it away. Someday if the logic changes, we won't have to change it at the other places.

Reply:
Makes sense. :+1: